server/src/main/java/org/elasticsearch/plugins/PluginsUtils.java

```diff
@@ -77,43 +79,110 @@ public static List<Path> findPluginDirs(final Path rootPath) throws IOException
      * Verify the given plugin is compatible with the current Elasticsearch installation.
      */
     public static void verifyCompatibility(PluginDescriptor info) {
-        if (info.isStable()) {
-            if (info.getElasticsearchVersion().major != Version.CURRENT.major) {
-                throw new IllegalArgumentException(
-                    "Stable Plugin ["
-                        + info.getName()
-                        + "] was built for Elasticsearch major version "
-                        + info.getElasticsearchVersion().major
-                        + " but version "
-                        + Version.CURRENT
-                        + " is running"
+        Matcher buildVersionMatcher = SemanticVersion.semanticPattern.matcher(Build.current().version());
+        // If we're not on a semantic version, assume plugins are compatible
```

Comment:
I'm not sure if we can -- if I understand things correctly, `PluginDescriptor#getElasticsearchVersion` will return a semantic version even if `Build.current().version()` is a commit hash.

Reply:
> PluginDescriptor#getElasticsearchVersion will return a semantic version

For serverless, official plugins/modules are built with whatever version identifier serverless uses. Right now that is a semantic version, it might become the commit hash in the future. But it shouldn't matter. If it's not a stable plugin (which only applies to non-serverless right now), the version can be opaque because we only care about equality of the plugin ES version with the running ES version.

Reply:
> the version can be opaque because we only care about equality of the plugin ES version with the running ES version.

Makes sense. I'll switch it.

Reply:
The failure from the serverless build has to do with the switch. PluginDescriptor is returning `8.12.0` from its info file, but current version is reported as a build hash. Do you think I should change things downstream, or revert? (And if I change things downstream, I think I'll need to revisit the work in https://github.com/elastic/elasticsearch/pull/100713#discussion_r1358927861 )
```
1> [2023-10-24T10:17:46,563][ERROR][o.e.b.Elasticsearch      ] [index-uSaWETn] fatal exception while booting Elasticsearch java.lang.IllegalArgumentException: Plugin [repository-url] was built for Elasticsearch version 8.12.0 but version 39ce28f8ad95728c14a28a3cda6dcadc88689e8f is running
1>    at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.plugins.PluginsUtils.verifyCompatibility(PluginsUtils.java:150)
1>    at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.plugins.PluginsService.loadBundle(PluginsService.java:462)
1>    at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.plugins.PluginsService.loadBundles(PluginsService.java:304)
1>    at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.plugins.PluginsService.<init>(PluginsService.java:174)
1>    at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.node.NodeServiceProvider.newPluginService(NodeServiceProvider.java:54)
1>    at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.node.NodeConstruction.construct(NodeConstruction.java:417)
1>    at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.node.NodeConstruction.prepareConstruction(NodeConstruction.java:244)
1>    at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.node.Node.<init>(Node.java:181)
1>    at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch$2.<init>(Elasticsearch.java:236)
1>    at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.initPhase3(Elasticsearch.java:236)
1>    at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:73)
```

Reply:
Ah, we *do* replace the version with a hash for Build, but not the version used in the actual build. Sorry for leading you down the wrong path here.

@mark-vieira How difficult would it be to swap out the version we use when building the composite project in serverless?

Reply:
Which "version" are we referring to here?

Reply:
The `elasticsearch` key in version.properties. Barring that, if we could replace the gradle project version on each project of the composite build? That is what plugin properties defaults to when building each plugin/module.

Reply:
That's probably not going to be easy to do. There's likely loads of logic that assumes that to be a semver string.

Reply:
I'm trying to follow along here -- is it this line under `BasePluginBuildPlugin#createBundleTasks` that determines what version gets written in the plugin properties file?

```
task.getElasticsearchVersion().set(Version.fromString(VersionProperties.getElasticsearch()).toString());
```

So we'd be looking to keep that as it is on stateful ES builds, but have it match the same hash that shows up in `org.elasticsearch.Build` on stateless builds of plugins. And `ServerlessBuildExtension` pulls that hash out of the `Change` entry in the Elasticsearch jar manifest, and that entry is configured by `ElasticsearchJavaPlugin#configureJarManifest`:
```
manifestPlugin.add("Change", toStringable(BuildParams::getGitRevision));
```
...and that's ultimately `GitInfo#revision`?

It seems like we should be able to write something different to the plugin properties, but I'm a little lost in the interdependencies right now. I can take another run at it tomorrow.

Reply:
I'll write a JIRA ticket for adjusting BasePluginBuildPlugin to write the string we're looking for into PluginDescriptor.